js/src/specs.js

```diff
@@ -4,155 +4,175 @@
     return new Dataset(this.database.database(), this.name);
   }
 
-  // Returns the value at the HEAD of this dataset, if any, or null otherwise.
-  value(): Promise<any> {
-    // Hm. Calling dataset() creates a Database that we then toss into the ether, which means we
-    // can't call close() on it. Ideally, we'd fix that.
-    return this.dataset().head()
-      .then(commit => commit && commit.value);
+  /**
+   * Returns a tuple of the database backed by this dataset, and value at the HEAD of this
+   * dataset. If this dataset doesn't have a HEAD, the value will be `null`.
+   *
+   * The caller should always call `close()` when done.
+   */
+  value(): Promise<[Database, Value|null]> {
+    const db = this.database.database();
+    return this.dataset().head().then(commit => [db, commit ? commit.value : null]);
   }
 }
 
-
-// A parsed specification for the location of a Noms hash.
-// For example: 'mem:sha1-5ba4be791d336d3184be7ee7dc598037f410ef96' or
-// 'https://ds.noms.io/aa/music:sha1-3ff6ee6add3490621a8886608cc8423dba3cf7ca'
-//
-// See "spelling objects" for details on supported syntaxes:
-// https://docs.google.com/document/d/1QgKcRS304llwU0ECahKtn8lGBFmT5zXzWr-5tah1S_4/edit
-export class HashSpec {
+/**
+ * A path to a Noms value within a database.
+ *
+ * E.g. the entirety of a spec like `http://demo.noms.io::foo.bar` or
+ * `http://demo.noms.io::#abcdef.bar`.
+ */
+export class PathSpec {
   database: DatabaseSpec;
-  hash: Hash;
-
-  // Returns a parsed spec, or null if the spec was invalid.
-  static parse(spec: string): ?HashSpec {
-    const match = spec.match(/^(.+)\:\:([\-sh0-9a-fA-F]+)$/);
-    if (!match) {
-      return null;
-    }
-
-    const [, dbSpec, hashPart] = match;
-    const hash = Hash.parse(hashPart);
-    if (!hash) {
-      return null;
-    }
-
-    const database = DatabaseSpec.parse(dbSpec);
-    if (!database) {
-      return null;
-    }
-
-    return new this(database, hash);
+  path: AbsolutePath;
+
+  /**
+   * Parses `str` as a PathSpec. Throws a SyntaxError if `str` isn't valid.
+   */
+  static parse(str: string): PathSpec {
+    const [database, pathStr] = splitAndParseDatabaseSpec(str);
+    const path = AbsolutePath.parse(pathStr);
+    return new PathSpec(database, path);
   }
 
-  constructor(database: DatabaseSpec, hash: Hash) {
+  constructor(database: DatabaseSpec, path: AbsolutePath) {
     this.database = database;
-    this.hash = hash;
+    this.path = path;
   }
 
-  // Returns the value for the spec'd reference, if any, or null otherwise.
-  value(): Promise<any> {
-    const database = this.database.database();
-    return database.readValue(this.hash).then(v => database.close().then(() => v));
+  /**
+   * Resolves this PathSpec, and returns the database it was resolved in, and the value it
+   * resolved to. If the value wasn't found, it will be `null`.
+   *
+   * The caller should always call `close()` when done.
+   */
+  value(): Promise<[Database, Value|null]> {
```

Comment:
I still think we should use spaces around `|` since all other binops require it.